src/txn/txn.c

```diff
@@ -1038,8 +1038,38 @@ __txn_fixup_prepared_update(
         hs_cursor->set_value(hs_cursor, &tw, tw.durable_stop_ts, tw.durable_start_ts,
           (uint64_t)WT_UPDATE_STANDARD, &hs_value);
         WT_ERR(hs_cursor->update(hs_cursor));
-    } else
-        WT_ERR(hs_cursor->remove(hs_cursor));
+    } else {
+        /*
+         * Remove the history store entry if checkpoint is not running. Otherwise, fix the history
+         * store entry with a prepared tombstone when the history store entry doesn't have its own
+         * stop timestamp.
+         */
+        if (txn_global->checkpoint_running) {
+            /* Don't update the history store entry if the entry already have a stop timestamp. */
+            if (fix_upd->type != WT_UPDATE_TOMBSTONE) {
+                tw.durable_stop_ts = fix_upd->durable_ts;
+                tw.stop_ts = fix_upd->start_ts;
+
+                /*
+                 * Use checkpoint next transaction id, so that this tombstone is not visible to the
+                 * rollback to stable and also checkpoint garbage collection cannot clean it as it
+                 * is greater than the global visible transaction id.
+                 */
+                tw.stop_txn = txn_global->checkpoint_next_txn_id;
```

Comment:
Can we race here? It is possible that after we checked the checkpoint_running flag, don't have a checkpoint_next_txn_id because it is allocated after we get the snapshot but we set the checkpoint_running flag earlier. It is also possible that we have entered the next checkpoint and the checkpoint_next_txn_id still points to the previous value, which is wrong.

Reply:
This is ok but we need to explain with a comment.